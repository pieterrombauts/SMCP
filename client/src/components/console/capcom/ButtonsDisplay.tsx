import React, {useState, useEffect} from 'react';
import styled from 'styled-components';
import moment from 'moment';

interface ButtonsProps {
  className?: string;
  time: string;
};

const UButtonsDisplay: React.FC<ButtonsProps> = ( props ) => {
  const [ signal, setSignal ] = useState<boolean>(true)
  const [ losTimes, setLosTimes ] = useState<string[]>(['2019-11-15T15:17:37', '2019-11-15T15:57:51', '2019-11-15T16:03:53', '2019-11-15T16:07:03', '2019-11-15T16:10:10', '2019-11-15T16:13:24','2019-11-15T16:52:26','2019-11-15T17:37:00'])
  const [ aosTimes, setAosTimes ] = useState<string[]>(['2019-11-15T15:17:52', '2019-11-15T15:59:28', '2019-11-15T16:05:46', '2019-11-15T16:09:19', '2019-11-15T16:12:07', '2019-11-15T16:14:25','2019-11-15T17:00:00','2019-11-15T17:44:00'])
  const [ nextAOS, setNextAOS ] = useState<string>("");
  const [ nextLOS, setNextLOS ] = useState<string>("");
  useEffect(() => {
    if (moment(props.time).isBefore(moment(aosTimes[0]))) {
      var hoursAOS = moment(aosTimes[0]).diff(moment(props.time), 'hours')
      var minutesAOS = moment(aosTimes[0]).diff(moment(props.time), 'minutes')
      var secondsAOS = moment(aosTimes[0]).diff(moment(props.time), 'seconds')
      setNextAOS(hoursAOS.toString().padStart(2,'0') + ":" + (minutesAOS - 60*hoursAOS).toString().padStart(2,'0') + ":" + (secondsAOS - 60*minutesAOS).toString().padStart(2,'0'));
    } else if (moment(props.time).isSameOrAfter(moment(aosTimes[0]))) {
      setSignal(true);
      var newAosTimes = aosTimes;
      newAosTimes.shift();
      setAosTimes(newAosTimes);
    }
    if (moment(props.time).isBefore(moment(losTimes[0]))) {
      var hoursLOS = moment(losTimes[0]).diff(moment(props.time), 'hours')
      var minutesLOS = moment(losTimes[0]).diff(moment(props.time), 'minutes')
      var secondsLOS = moment(losTimes[0]).diff(moment(props.time), 'seconds')
      setNextLOS(hoursLOS.toString().padStart(2,'0') + ":" + (minutesLOS - 60*hoursLOS).toString().padStart(2,'0') + ":" + (secondsLOS - 60*minutesLOS).toString().padStart(2,'0'));
    } else if (moment(props.time).isSameOrAfter(moment(losTimes[0]))) {
      setSignal(false);
      var newLosTimes = losTimes;
      newLosTimes.shift();
      setLosTimes(newLosTimes);
    }
  }, [props.time, signal, losTimes, aosTimes])
  return(
    <div className={props.className}>
      <svg preserveAspectRatio='xMidYMid meet' viewBox='0 0 1300 990'>
        <rect x="740" y="20" rx="5" ry="5" width="560" height="400" style={{fill:"#393939", stroke:"#393939", strokeWidth:5, opacity:0.5}} />
        

        
        <rect x="770" y="40" rx="5" ry="5" width="200" height="100" style={{fill: signal ? "#00CC00" : "#004000", stroke:"#393939", strokeWidth:5}} />
        
        <text x="810" y="110" fill={signal ? "white" : "grey"} fontSize="50" fontFamily="Verdana">AOS</text>

        <rect x="1060" y="40" rx="5" ry="5" width="200" height="100" style={{fill: signal ? "#640000" : "#FF0000", stroke:"#393939", strokeWidth:5}} />
        <text x="1110" y="110" fill={signal ? "grey" : "white"} fontSize="50" fontFamily="Verdana">LOS</text>

        <text x="780" y="230" fill="white" fontSize="50" fontFamily="Verdana">Next AOS {nextAOS}</text>
        <text x="780" y="310" fill="white" fontSize="50" fontFamily="Verdana">Next LOS {nextLOS}</text>

        <rect x='0' y='20' rx="5" ry="5" width="700" height="400" style={{fill:"#333333", stroke:"#393939", strokeWidth:5, opacity:0.5}} />
        
        <text x="30" y="100" fill="white" fontSize="50" fontFamily="Verdana" text-decoration="underline">Caution & Warning </text>
        
        <rect x="20" y="140" rx="5" ry="5" width="110" height="90" style={{fill:"white", stroke:"#393939", strokeWidth:5}} />
        <text x="40" y="195" fill="black" fontSize="30" fontFamily="Verdana">FIRE</text>

        <rect x="160" y="140" rx="5" ry="5" width="110" height="90" style={{fill:"white", stroke:"#393939", strokeWidth:5}} />
        <text x="180" y="195" fill="black" fontSize="30" fontFamily="Verdana">Î”P</text>

        <rect x="300" y="140" rx="5" ry="5" width="110" height="90" style={{fill:"white", stroke:"#393939", strokeWidth:5}} />
        <text x="320" y="195" fill="black" fontSize="30" fontFamily="Verdana">ATM</text>

        <rect x="440" y="140" rx="5" ry="5" width="110" height="90" style={{fill:"white", stroke:"#393939", strokeWidth:5}} />
        <text x="450" y="195" fill="black" fontSize="18" fontFamily="Verdana">WARNING</text>

        <rect x="580" y="140" rx="5" ry="5" width="110" height="90" style={{fill:"white", stroke:"#393939", strokeWidth:5}} />
        <text x="590" y="195" fill="black" fontSize="18" fontFamily="Verdana">CAUTION</text>

        <rect x="580" y="300" rx="5" ry="5" width="110" height="90" style={{fill:"white", stroke:"#393939", strokeWidth:5}} />
        <text x="595" y="355" fill="black" fontSize="30" fontFamily="Verdana">TEST</text>
      </svg>
    </div>
  )
};

const ButtonsDisplay = styled(UButtonsDisplay)`
  svg {
    height: 100%;
    position: absolute;
    left: 0;
  }
`;

export default ButtonsDisplay;