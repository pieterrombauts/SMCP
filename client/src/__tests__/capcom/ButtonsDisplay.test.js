import React from "react";
import 'jest-styled-components'
import toJson from 'enzyme-to-json'
import ButtonsDisplay from '../../components/console/capcom/ButtonsDisplay'
import { shallow, mount, render} from 'enzyme';

describe("Buttons display display", () => {
  it('renders correctly', () => {
    const wrapper = mount(<ButtonsDisplay time={'2019-11-19T15:00:00'} />)
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  })
})