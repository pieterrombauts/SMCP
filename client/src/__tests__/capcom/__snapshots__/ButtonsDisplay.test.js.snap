// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttons display display renders correctly 1`] = `
.c0 svg {
  height: 100%;
  position: absolute;
  left: 50%;
  -webkit-transform: translate(-50%,0%);
  -ms-transform: translate(-50%,0%);
  transform: translate(-50%,0%);
}

<Styled(UButtonsDisplay)
  time="2019-11-19T15:00:00"
>
  <UButtonsDisplay
    className="c0"
    time="2019-11-19T15:00:00"
  >
    <div
      className="c0"
    >
      <svg
        preserveAspectRatio="xMidYMid meet"
        viewBox="0 0 1300 990"
      >
        <rect
          height="400"
          rx="5"
          ry="5"
          style={
            Object {
              "fill": "#393939",
              "opacity": 0.5,
              "stroke": "#393939",
              "strokeWidth": 5,
            }
          }
          width="560"
          x="740"
          y="20"
        />
        <rect
          height="100"
          rx="5"
          ry="5"
          style={
            Object {
              "fill": "#004000",
              "stroke": "#393939",
              "strokeWidth": 5,
            }
          }
          width="200"
          x="770"
          y="40"
        />
        <text
          fill="grey"
          fontFamily="Verdana"
          fontSize="50"
          x="810"
          y="110"
        >
          AOS
        </text>
        <rect
          height="100"
          rx="5"
          ry="5"
          style={
            Object {
              "fill": "#FF0000",
              "stroke": "#393939",
              "strokeWidth": 5,
            }
          }
          width="200"
          x="1060"
          y="40"
        />
        <text
          fill="white"
          fontFamily="Verdana"
          fontSize="50"
          x="1110"
          y="110"
        >
          LOS
        </text>
        <text
          fill="white"
          fontFamily="Verdana"
          fontSize="50"
          x="780"
          y="230"
        >
          Next AOS 
        </text>
        <text
          fill="white"
          fontFamily="Verdana"
          fontSize="50"
          x="780"
          y="310"
        >
          Next LOS 
        </text>
        <rect
          height="400"
          rx="5"
          ry="5"
          style={
            Object {
              "fill": "#333333",
              "opacity": 0.5,
              "stroke": "#393939",
              "strokeWidth": 5,
            }
          }
          width="700"
          x="0"
          y="20"
        />
        <text
          fill="white"
          fontFamily="Verdana"
          fontSize="50"
          textDecoration="underline"
          x="30"
          y="100"
        >
          Caution & Warning 
        </text>
        <rect
          height="90"
          rx="5"
          ry="5"
          style={
            Object {
              "fill": "white",
              "stroke": "#393939",
              "strokeWidth": 5,
            }
          }
          width="110"
          x="20"
          y="140"
        />
        <text
          fill="black"
          fontFamily="Verdana"
          fontSize="30"
          x="40"
          y="195"
        >
          FIRE
        </text>
        <rect
          height="90"
          rx="5"
          ry="5"
          style={
            Object {
              "fill": "white",
              "stroke": "#393939",
              "strokeWidth": 5,
            }
          }
          width="110"
          x="160"
          y="140"
        />
        <text
          fill="black"
          fontFamily="Verdana"
          fontSize="30"
          x="180"
          y="195"
        >
          Î”P
        </text>
        <rect
          height="90"
          rx="5"
          ry="5"
          style={
            Object {
              "fill": "white",
              "stroke": "#393939",
              "strokeWidth": 5,
            }
          }
          width="110"
          x="300"
          y="140"
        />
        <text
          fill="black"
          fontFamily="Verdana"
          fontSize="30"
          x="320"
          y="195"
        >
          ATM
        </text>
        <rect
          height="90"
          rx="5"
          ry="5"
          style={
            Object {
              "fill": "#F54242",
              "stroke": "#393939",
              "strokeWidth": 5,
            }
          }
          width="110"
          x="440"
          y="140"
        />
        <text
          fill="white"
          fontFamily="Verdana"
          fontSize="18"
          x="450"
          y="195"
        >
          WARNING
        </text>
        <rect
          height="90"
          rx="5"
          ry="5"
          style={
            Object {
              "fill": "white",
              "stroke": "#393939",
              "strokeWidth": 5,
            }
          }
          width="110"
          x="580"
          y="140"
        />
        <text
          fill="black"
          fontFamily="Verdana"
          fontSize="18"
          x="590"
          y="195"
        >
          CAUTION
        </text>
        <rect
          height="90"
          rx="5"
          ry="5"
          style={
            Object {
              "fill": "white",
              "stroke": "#393939",
              "strokeWidth": 5,
            }
          }
          width="110"
          x="580"
          y="300"
        />
        <text
          fill="black"
          fontFamily="Verdana"
          fontSize="30"
          x="595"
          y="355"
        >
          TEST
        </text>
      </svg>
    </div>
  </UButtonsDisplay>
</Styled(UButtonsDisplay)>
`;
